src/gfx/renderer_cache.hpp

```diff
@@ -2,23 +2,86 @@
 #define A2997EDF_4E03_48F4_AF46_10E0BEE1DFE3
 
 #include "renderer_types.hpp"
+#include "pipeline_hash_collector.hpp"
 #include "hasherxxh128.hpp"
+#include <magic_enum.hpp>
 #include <unordered_map>
 
 namespace gfx::detail {
-struct PipelineCache {
-  std::unordered_map<Hash128, CachedPipelinePtr> map;
 
-  void clear() { map.clear(); }
+struct PipelineHashCache final : public IPipelineHashStorage {
+  struct Entry {
+    Hash128 hash;
+  };
+  std::unordered_map<UniqueId, Entry> perType[magic_enum::enum_count<UniqueIdTag>()];
+
+  const Hash128 *find(UniqueId id) {
```

Comment:
I'm confused, why a 64 bit preudo-unique id is used to fetch a 128 bits hash, which could be the ID itself ?
The uniqueid class is also weird. Why we need that? I feel the work with ENTT might help maybe.

Reply:
This was an optimization because of the way I hash things, for example:

A drawable can have a Material that modifies it's pipeline hash, it could also have some Features attached that modify it. The material could also have features attached that modify it.

Because there might be many objects that reference the same material for example, whenever such a reference is encountered it is stored like this so it doesn't have to be computed multiple times.

Reply:
Technically there's some similarities with entt, if we're using it anyway I might try using it for the rendering where applicable.

Reply:
Yeah that's the plan. I guess let's keep things this way for now. But I doubt the uniqueid class is really "unique" 🤪

Reply:
Not random or anything but definitely unique during application lifetime every time one is generated.
As long as the uint64_t doesn't wrap, which as en example would take about 4998 years of creating 1 million unique objects every frame at 120 FPS 😄

Reply:
haha plenty